### ISSUE TITLE
Very large amounts of memory allocated during training

### ISSUE THREAD

**Issue by Macbor811** [jdermody/brightwire#41]
Hello. I'm trying to implement a pretty simple neural network using this library. The data I'm using for training is around 2 million rows of 128-element float vectors, each row with a label. It takes around 6.5 GB of RAM before training. 
Unfortunately, during training it starts growing rapidly, reaching above 14 GB just after 2 epochs. As my machine has only 16 GB total, its not surprising that it slows down everything dramatically while operating system attempts to find more and more space to fit it all. 

Here's the code:
`

            var context = new BrightDataContext();
            context.LinearAlgebraProvider = context.UseNumericsLinearAlgebra();

            var dataTableTrain = CreateDataTable(context, inputsTrain, labelsTrain, labelNamesTrain);
            var dataTableTest = CreateDataTable(context, inputsTest, labelsTest, labelNamesTest);

            var graph = context.CreateGraphFactory();

            graph.CurrentPropertySet
                .Use(graph.GradientDescent.Adam)
                .Use(graph.GaussianWeightInitialisation(false, 0.1f, GaussianVarianceCalibration.SquareRoot2N));

            var trainingData = graph.CreateDataSource(dataTableTrain);

            var engine = graph.CreateTrainingEngine(trainingData, graph.ErrorMetric.OneHotEncoding);
            uint numIterations = 15;

            graph.Connect(engine)
                .AddFeedForward(trainingData.InputSize)
                .Add(graph.ReluActivation())
                .AddFeedForward(trainingData.GetOutputSizeOrThrow())
                .Add(graph.SoftMaxActivation())
                .AddBackpropagation();
            
            engine.Train(numIterations, trainingData.CloneWith(dataTableTest));
`

I don't have a lot of experience with ML, but so far I've tried Encog and Tensorflow libraries and the memory consumption during training stayed pretty much constant with them, so something has to be wrong here.
According to Dynamic Program Analysis from Rider IDE, the biggest culprit seems to be:

`
at LearningContext.Update(List)
at TrainingEngine.Train(IGraphExecutionContext, Action)
at ExtensionMethods.Train(IGraphTrainingEngine, uint, IDataSource, Action, int)
at ProgramBrightwire.Main(String[])
`

but there are many, many other places branching from Train method where large amounts of memory are allocated, too.

**Comment by jdermody** [comment]
Thanks for raising this - it looks like there was a missing call to release memory in the base level memory context

**Comment by jdermody** [comment]
Fixed in 3.0.1
